### PR TITLE
Ksagiyam/periodic extrusion

### PR DESCRIPTION
**Pull request by ksagiyam**
Closes https://github.com/firedrakeproject/firedrake/discussions/2570

Enable periodic extrusion.

- ~~Need to do more testing,~~ Done.
- ~~Need to fix I/O for this new mesh.~~ Done.

~~Depends on: https://github.com/OP2/PyOP2/pull/675~~ (Merged)

### CODE REVIEW

File: tests/extrusion/test_extruded_periodic.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    L = inner(f, v) * dx
+    exact = Function(V).interpolate(cos(2 * pi * x) * cos(2 * pi * y))
+    bc = DirichletBC(V, exact, (1, 2))
+    sol = Function(V)
+    solve(a == L, sol, bcs=[bc])
+    assert sqrt(assemble(inner(sol - exact, sol - exact) * dx)) < 1.e-7
+
+
+@pytest.mark.parallel(nprocs=3)
+def test_extruded_periodic_annulus():
```
Comment (JDBetteridge) [inline tests/extrusion/test_extruded_periodic.py:115]: Could do the same for a torus mesh

Comment (ksagiyam) [reply tests/extrusion/test_extruded_periodic.py:115]: Added a comment in `ExtrudedMesh` doc.

File: firedrake/mesh.py [hunk 2]
```diff
@@ -2046,7 +2055,7 @@ def Mesh(meshfile, **kwargs):
 
 
 @PETSc.Log.EventDecorator("CreateExtMesh")
-def ExtrudedMesh(mesh, layers, layer_height=None, extrusion_type='uniform', kernel=None, gdim=None, name=None):
+def ExtrudedMesh(mesh, layers, layer_height=None, extrusion_type='uniform', periodic=False, kernel=None, gdim=None, name=None):
```
Comment (connorjward) [inline firedrake/mesh.py:2058]: Would be useful to have `TorusMesh` and `AnnulusMesh` functions in `utility_meshes.py`.

**Review by connorjward — CHANGES_REQUESTED** [review]

File: firedrake/mesh.py (hunk 2 above)
Comment (ksagiyam) [reply firedrake/mesh.py:2164]: Added `AnnulusMesh` and `SolidTorusMesh`.

File: firedrake/utility_meshes.py [hunk 3]
```diff
@@ -2540,6 +2542,110 @@ def TorusMesh(
...
+    r,
+    nr=4,
+    nt=32,
+    distribution_parameters=None,
+    comm=COMM_WORLD,
+    name=mesh.DEFAULT_MESH_NAME,
+    distribution_name=None,
+    permutation_name=None,
+):
+    """Generate a solid toroidal mesh (with axis z) periodically extruding a disk mesh
```
Comment (dham) [inline firedrake/utility_meshes.py:2557]: Wrong docstring.

File: firedrake/utility_meshes.py [hunk 4]
```diff
@@ -2540,6 +2542,110 @@ def TorusMesh(
     return m
 
 
+@PETSc.Log.EventDecorator()
+def AnnulusMesh(
+    R,
+    r,
+    nr=4,
+    nt=32,
+    distribution_parameters=None,
```
Comment (dham) [inline firedrake/utility_meshes.py:2551]: Not in docstring.

File: firedrake/utility_meshes.py [hunk 5]
```diff
@@ -2540,6 +2542,110 @@ def TorusMesh(
...
+    comm=COMM_WORLD,
+    name=mesh.DEFAULT_MESH_NAME,
+    distribution_name=None,
+    permutation_name=None,
+):
+    """Generate a solid toroidal mesh (with axis z) periodically extruding a disk mesh
+
+    :arg R: The major radius
+    :arg r: The minor radius
+    :kwarg nR: (optional), number of cells in the major direction (min 3, defaults to 8)
```
Comment (dham) [inline firedrake/utility_meshes.py:2611]: Don't put default in docstring.

**Review by dham — CHANGES_REQUESTED** [review]
Just fix docstrings.

File: firedrake/utility_meshes.py (hunk 3 above)
Comment (ksagiyam) [reply firedrake/utility_meshes.py]: Fixed.

File: firedrake/utility_meshes.py (hunk 4 above)
Comment (ksagiyam) [reply firedrake/utility_meshes.py:2573]: Fixed.

File: firedrake/utility_meshes.py (hunk 5 above)
Comment (ksagiyam) [reply firedrake/utility_meshes.py]: Fixed.

**Review by dham — APPROVED** [review]

**Pull request merged**